[Lens] close dimension editor when editing color palette and clicking outside

mbondyra opened the pull request:
https://github.com/elastic/kibana/assets/4283304/e85b7f97-114a-4b16-8b72-02eee5a121b9

Fixes https://github.com/elastic/kibana/issues/176247 (I removed the flaky tests because they were bad. Replaced them with functional ones, flaky test runner here: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/5071)

Fixes an issue with the dimension editor visible on the video. 
When the second flyout (palette) is open and we change the chart (for example from Bar Vertical Stacked to Bar Vertical when only having a breakdown dimension defined), the breakdown dimension moves to horizontal axis, but it’s not reflected on the flyout (we still get the settings for breakdown, palette etc). I initially fixed it by closing two flyouts and not fixing updating the state of the flyout, but then I found a solution to also update the state instead.

I did a refactor here for the palette flyout and generalized it to setting with sibling flyout component. I also moved some repeating logic inside of it. 
Styles are moved, but nothing changes from the user perspective.  

The only visual change is to unify palette settings by removing `Edit` copy from the button (some palette settings had it, other didn't). We can also do it the other way around if it makes more sense:

before:
<img width="422" alt="Screenshot 2024-01-29 at 11 17 47" src="https://github.com/elastic/kibana/assets/4283304/60499df6-60ea-4e81-b087-bd20d5bd0940">

after:
<img width="422" alt="Screenshot 2024-01-29 at 11 18 32" src="https://github.com/elastic/kibana/assets/4283304/2bc29ec2-4aad-4662-bef0-ea3fe06c50e1">

mbondyra commented:
/ci

mbondyra commented:
/ci

mbondyra commented:
/ci

mbondyra commented:
/ci

mbondyra commented:
/ci

mbondyra commented:
/ci

mbondyra commented:
/ci

elasticmachine commented:
Pinging @elastic/kibana-visualizations (Team:Visualizations)

MichaelMarcialis commented on x-pack/plugins/lens/public/shared_components/coloring/palette_panel_container.tsx:
```diff
@@ -5,105 +5,45 @@
+            />
+          </EuiFlexItem>
+          <EuiFlexItem grow={false}>
+            <EuiButtonIcon
+              data-test-subj="lns_colorEditing_trigger"
+              aria-label={i18n.translate('xpack.lens.colorMapping.editColorMappingButton', {
+                defaultMessage: 'Edit palette',
+              })}
+              iconType="controlsHorizontal"
+              onClick={onClick}
+              size="xs"
+            />
```
Would it be possible to add a tooltip to this button and have it's text (and the existing `aria-label`) be "Edit colors"?

mbondyra replied:
Done!

MichaelMarcialis reviewed (commented):
Thanks for putting this together, @mbondyra. In my testing, clicking outside a second-level flyout (such as the color configuration flyout) still doesn't close all the flyouts. Only the color one is being closed on the first click (as shown in your video example). The original flyout still requires a second click to close.

I also left one additional request below. Thanks!

mbondyra commented:
Hi @MichaelMarcialis thank you for taking a look! I added a tooltip. Regarding the main problem, I updated the description. I am sorry I didn't mention it before, we decided to keep the flyout open.

dej611 reviewed (approved):
Had another go and tested it the new behaviour 👍 
Nice all that code removals.

MichaelMarcialis commented on x-pack/plugins/lens/public/shared_components/setting_with_sibling_flyout.tsx:
```diff
@@ -0,0 +1,120 @@
+                className="lnsSettingWithSiblingFlyout"
+              >
+                <EuiFlyoutHeader hasBorder className="lnsSettingWithSiblingFlyout__header">
+                  <EuiFlexGroup gutterSize="s" alignItems="center">
```
I think this `EuiFlexGroup` needs a `responsive={false}` prop to prevent breaking at small viewport sizes.

```suggestion
                  <EuiFlexGroup gutterSize="s" alignItems="center" responsive={false}>
```

MichaelMarcialis reviewed (approved):
Thanks, @mbondyra. I left one small comment below, but nothing worth holding you up over. Assuming you can address it, this looks good to me.

Additionally, I'm not sure this is related to this specific PR, but I noticed that the edit color flyout is stacking with the original flyout instead of covering it at smaller viewport sizes. If not related to this PR, let me know and I'll open a bug issue.

![CleanShot 2024-02-12 at 12 32 52](https://github.com/elastic/kibana/assets/3884767/cb63c317-52aa-43e3-9a09-3f6152716e0a)

mbondyra commented:
Thank you @MichaelMarcialis! The bug you mentioned is also reproducible on main so I'd appreciate an issue when you have  a moment 🙏🏼
